be/src/exec/spill/spilled_stream.cpp

```diff
@@ -36,16 +36,267 @@
+        if (res.status().is_end_of_file()) {
+            _readable.reset();
+            TRACE_SPILL_LOG << "read chunk size" << _output_chunks - 1;
+            _output_chunks = 0;
+        }
+
+        RETURN_IF(!res.status().is_end_of_file(), res);
+    }
+    return nullptr;
+}
+
+void SequentialFileStream::close() {
+    _readable.reset();
+}
+
+class BufferedSpilledStream final : public SpilledInputStream {
+public:
+    BufferedSpilledStream(int capacity, std::shared_ptr<SpilledInputStream> stream)
+            : _capacity(capacity), _raw_input_stream(std::move(stream)) {}
+    ~BufferedSpilledStream() override = default;
+
+    bool buffer_fulled() { return _chunk_buffer.get_size() == _capacity || eof(); }
+    bool has_chunk() { return _chunk_buffer.get_size() > 0 || eof(); }
+
+    bool is_ready() override { return has_chunk(); }
+
+    void close() override {}
+
+    bool acquire() {
+        bool expected = false;
+        return _is_running.compare_exchange_strong(expected, true);
+    }
+
+    void release() { _is_running = false; }
+
+    StatusOr<ChunkUniquePtr> read(SpillFormatContext& context) override;
+
+    // read chunk from buffer
+    StatusOr<ChunkUniquePtr> read_from_buffer();
+
+    // read upstream chunk and move it to inner buffer
+    Status read_to_buffer(SpillFormatContext& context);
+
+private:
+    int _capacity;
+    UnboundedBlockingQueue<ChunkUniquePtr> _chunk_buffer;
+    std::shared_ptr<SpilledInputStream> _raw_input_stream;
+    std::atomic_bool _is_running{};
+    size_t _read_rows_from_buffer = 0;
+};
+
+StatusOr<ChunkUniquePtr> BufferedSpilledStream::read_from_buffer() {
+    if (_chunk_buffer.empty()) {
+        CHECK(eof());
+        return Status::EndOfFile("eos");
+    }
+    ChunkUniquePtr res;
+    CHECK(_chunk_buffer.try_get(&res));
+    _read_rows_from_buffer += res->num_rows();
+
+    return res;
+}
+
+StatusOr<ChunkUniquePtr> BufferedSpilledStream::read(SpillFormatContext& context) {
+    if (has_chunk()) {
+        return read_from_buffer();
+    }
+    return _raw_input_stream->read(context);
+}
+
+Status BufferedSpilledStream::read_to_buffer(SpillFormatContext& context) {
+    DCHECK(!buffer_fulled());
+
+    // get chunk from stream
+    auto res = _raw_input_stream->read(context);
+    if (res.ok()) {
+        // put it to chunk buffer
+        _chunk_buffer.put(std::move(res.value()));
+        return Status::OK();
+    } else if (res.status().is_end_of_file()) {
+        mark_eof();
+        return Status::OK();
+    } else {
+        return res.status();
+    }
+}
+
+class SortedFileStream final : public SpilledInputStream {
+public:
+    SortedFileStream(SpilledInputStreamList streams, RuntimeState* state);
+    ~SortedFileStream() override = default;
+
+    Status init(const SortExecExprs* exprs, const SortDescs* descs);
+    StatusOr<ChunkUniquePtr> read(SpillFormatContext& context) override;
+    bool is_ready() override { return _merger.is_data_ready(); }
```

Comment:
const?

Reply:
MergeCursorsCascade::is_data_ready is not const member function